File: src/package/deployer_main.cpp
#include "package_core.h"

#include <windows.h>
#include <shlobj.h>

#include <fcitx5_windows/version.h>

#include <array>
#include <filesystem>
#include <iostream>
#include <string>
#include <string_view>
#include <vector>

namespace {

constexpr std::uint64_t kMaximumArtifactBytes = 128ULL * 1024ULL * 1024ULL;

class Handle final {
 public:
  explicit Handle(HANDLE handle = INVALID_HANDLE_VALUE) : handle_(handle) {}
  ~Handle() {
    if (handle_ != INVALID_HANDLE_VALUE && handle_ != nullptr) CloseHandle(handle_);
  }
  Handle(const Handle&) = delete;
  Handle& operator=(const Handle&) = delete;
  [[nodiscard]] HANDLE get() const noexcept { return handle_; }

 private:
  HANDLE handle_;
};

bool is_elevated() {
  HANDLE raw_token = nullptr;
  if (!OpenProcessToken(GetCurrentProcess(), TOKEN_QUERY, &raw_token)) return false;
  Handle token(raw_token);
  TOKEN_ELEVATION elevation{};
  DWORD returned = 0;
  return GetTokenInformation(token.get(), TokenElevation, &elevation, sizeof(elevation),
                             &returned) != FALSE &&
         elevation.TokenIsElevated != 0U;
}

std::filesystem::path module_path() {
  std::wstring buffer(32'768U, L'\0');
  const DWORD length =
      GetModuleFileNameW(nullptr, buffer.data(), static_cast<DWORD>(buffer.size()));
  if (length == 0U || length >= buffer.size()) {
    throw fcitx::package::PackageError("privilege_boundary", "module path is unavailable");
  }
  buffer.resize(length);
  return std::filesystem::path(buffer);
}

bool ordinal_path_prefix(const std::filesystem::path& candidate,
                         const std::filesystem::path& parent) {
  const auto normalized_candidate = std::filesystem::weakly_canonical(candidate).wstring();
  auto normalized_parent = std::filesystem::weakly_canonical(parent).wstring();
  if (!normalized_parent.ends_with(L'\\')) normalized_parent.push_back(L'\\');
  if (normalized_candidate.size() <= normalized_parent.size()) return false;
  return CompareStringOrdinal(normalized_candidate.data(), static_cast<int>(normalized_parent.size()),
                              normalized_parent.data(), static_cast<int>(normalized_parent.size()),
                              TRUE) == CSTR_EQUAL;
}

std::filesystem::path protected_install_root() {
  const auto executable = module_path();
  const auto root = executable.parent_path().parent_path();
  wchar_t program_files[MAX_PATH]{};
  if (SHGetFolderPathW(nullptr, CSIDL_PROGRAM_FILES, nullptr, SHGFP_TYPE_CURRENT, program_files) !=
          S_OK ||
      !ordinal_path_prefix(root, program_files) ||
      _wcsicmp(executable.filename().c_str(), L"fcitx5-deployer.exe") != 0 ||
      _wcsicmp(executable.parent_path().filename().c_str(), L"bin") != 0) {
    throw fcitx::package::PackageError(
        "privilege_boundary", "deployer must run from Program Files/Fcitx5/bin");
  }
  return root;
}

void copy_exclusive_artifact(const std::filesystem::path& source,
                             const std::filesystem::path& destination) {
  Handle input(CreateFileW(source.c_str(), GENERIC_READ, 0, nullptr, OPEN_EXISTING,
                           FILE_ATTRIBUTE_NORMAL | FILE_FLAG_OPEN_REPARSE_POINT, nullptr));
  if (input.get() == INVALID_HANDLE_VALUE) {
    throw fcitx::package::PackageError("artifact_changed", "artifact cannot be opened exclusively");
  }
  FILE_ATTRIBUTE_TAG_INFO tag{};
  LARGE_INTEGER size{};
  if (!GetFileInformationByHandleEx(input.get(), FileAttributeTagInfo, &tag, sizeof(tag)) ||
      (tag.FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) != 0U ||
      !GetFileSizeEx(input.get(), &size) || size.QuadPart <= 0 ||
      static_cast<std::uint64_t>(size.QuadPart) > kMaximumArtifactBytes) {
    throw fcitx::package::PackageError("artifact_changed", "artifact identity or size is unsafe");
  }
  Handle output(CreateFileW(destination.c_str(), GENERIC_WRITE, 0, nullptr, CREATE_NEW,
                            FILE_ATTRIBUTE_NORMAL | FILE_FLAG_WRITE_THROUGH, nullptr));
  if (output.get() == INVALID_HANDLE_VALUE) {
    throw fcitx::package::PackageError("io_error", "protected artifact copy cannot be created");
  }
  std::vector<std::byte> buffer(64U * 1024U);
  for (;;) {
    DWORD read = 0;
    if (!ReadFile(input.get(), buffer.data(), static_cast<DWORD>(buffer.size()), &read, nullptr)) {
      throw fcitx::package::PackageError("io_error", "artifact read failed");
    }
    if (read == 0U) break;
    DWORD written = 0;
    if (!WriteFile(output.get(), buffer.data(), read, &written, nullptr) || written != read) {
      throw fcitx::package::PackageError("io_error", "protected artifact copy failed");
    }
  }
  if (!FlushFileBuffers(output.get())) {
    throw fcitx::package::PackageError("io_error", "protected artifact flush failed");
  }
}

bool is_hex_digest(std::string_view value) {
  if (value.size() != 64U) return false;
  for (const char character : value) {
    if (!((character >= '0' && character <= '9') ||
          (character >= 'a' && character <= 'f') ||
          (character >= 'A' && character <= 'F'))) {
      return false;
    }
  }
  return true;
}

void activate(const std::filesystem::path& source, std::string_view expected_hash,
              std::string_view transaction_id) {
  if (!is_elevated() || !is_hex_digest(expected_hash) ||
      !fcitx::package::is_safe_relative_package_path(transaction_id) ||
      transaction_id.find('/') != std::string_view::npos) {
    throw fcitx::package::PackageError("privilege_boundary", "deployer request is invalid");
  }
  const auto root = protected_install_root();
  const auto keyring = root / "security/trusted-keys.json";
  const auto keys = fcitx::package::read_trusted_keys(keyring);
  const auto transactions = root / ".transactions" / std::filesystem::path(transaction_id);
  if (std::filesystem::exists(transactions)) {
    throw fcitx::package::PackageError("transaction_exists", "protected transaction exists");
  }
  std::filesystem::create_directories(transactions);
  const auto protected_archive = transactions / "artifact.fcpkg";
  try {
    copy_exclusive_artifact(source, protected_archive);
    if (fcitx::package::hex_sha256(fcitx::package::sha256_file(protected_archive)) !=
        expected_hash) {
      throw fcitx::package::PackageError("hash_mismatch",
                                         "artifact changed across elevation boundary");
    }
    const auto staged = fcitx::package::stage_verified_archive(
        protected_archive, root, transaction_id, keys);
    fcitx::package::activate_staged_package(staged, root, keys);
    std::error_code ignored;
    std::filesystem::remove_all(transactions, ignored);
  } catch (...) {
    std::error_code ignored;
    std::filesystem::remove_all(transactions, ignored);
    throw;
  }
}

}  // namespace

int wmain(int argc, wchar_t** argv) {
  try {
    if (argc == 2 && std::wstring_view(argv[1]) == L"--version") {
      std::cout << "fcitx5-deployer " << fcitx::windows::version() << '\n';
      return 0;
    }
    if (argc == 2 && std::wstring_view(argv[1]) == L"--self-test") {
      try {
        static_cast<void>(protected_install_root());
        return 1;
      } catch (const fcitx::package::PackageError&) {
        return 0;
      }
    }
    if (argc == 5 && std::wstring_view(argv[1]) == L"--activate") {
      activate(argv[2], std::filesystem::path(argv[3]).string(),
               std::filesystem::path(argv[4]).string());
      return 0;
    }
    std::wcerr << L"Usage: fcitx5-deployer --activate LOCAL_ARCHIVE SHA256 TRANSACTION_ID\n";
    return 1;
  } catch (const fcitx::package::PackageError& error) {
    std::cerr << error.code() << ": " << error.what() << '\n';
    return 2;
  } catch (const std::exception& error) {
    std::cerr << "internal_error: " << error.what() << '\n';
    return 3;
  }
}

File: rust/package-core/src/deployer_main.rs
#![deny(unsafe_code)]
#![deny(unsafe_op_in_unsafe_fn)]

use std::ffi::OsString;
use std::path::{Path, PathBuf};

use fcitx5_package_core::{
    activate_staged_payload_tree, is_safe_relative_package_path, parse_trusted_keys, sha256_digest,
    stage_validated_archive_zip, TrustedKey,
};

const MAXIMUM_ARTIFACT_BYTES: u64 = 128 * 1024 * 1024;

fn version() -> &'static str {
    option_env!("FCITX_WINDOWS_VERSION").unwrap_or(env!("CARGO_PKG_VERSION"))
}

fn main() {
    let exit_code = match run(std::env::args_os().collect()) {
        Ok(code) => code,
        Err(DeployError::Package { code, message }) => {
            eprintln!("{code}: {message}");
            2
        }
        Err(DeployError::Internal(message)) => {
            eprintln!("internal_error: {message}");
            3
        }
    };
    std::process::exit(exit_code);
}

fn run(args: Vec<OsString>) -> Result<i32, DeployError> {
    if args.len() == 2 && args[1] == "--version" {
        println!("fcitx5-deployer {}", version());
        return Ok(0);
    }
    if args.len() == 2 && args[1] == "--self-test" {
        return match protected_install_root() {
            Ok(_) => Ok(1),
            Err(DeployError::Package { .. }) => Ok(0),
            Err(error) => Err(error),
        };
    }
    if args.len() == 5 && args[1] == "--activate" {
        activate(
            &PathBuf::from(&args[2]),
            &unicode_arg(&args[3])?,
            &unicode_arg(&args[4])?,
        )?;
        return Ok(0);
    }
    eprintln!("Usage: fcitx5-deployer --activate LOCAL_ARCHIVE SHA256 TRANSACTION_ID");
    Ok(1)
}

fn activate(source: &Path, expected_hash: &str, transaction_id: &str) -> Result<(), DeployError> {
    if !win32::is_elevated()
        || !is_hex_digest(expected_hash)
        || !is_safe_relative_package_path(transaction_id)
        || transaction_id.contains('/')
        || transaction_id.contains('\\')
    {
        return Err(package_error(
            "privilege_boundary",
            "deployer request is invalid",
        ));
    }
    let root = protected_install_root()?;
    let keys = read_trusted_keys(root.join("security").join("trusted-keys.json"))?;
    let transactions = root.join(".transactions").join(transaction_id);
    if transactions.exists() {
        return Err(package_error(
            "transaction_exists",
            "protected transaction exists",
        ));
    }
    std::fs::create_dir_all(&transactions).map_err(|error| {
        internal_error(format!("transaction directory creation failed: {error}"))
    })?;
    let protected_archive = transactions.join("artifact.fcpkg");
    let activated = (|| {
        win32::copy_exclusive_artifact(source, &protected_archive)?;
        let bytes = std::fs::read(&protected_archive)
            .map_err(|error| package_error("io_error", format!("artifact read failed: {error}")))?;
        if sha256_digest(&bytes).as_str() != expected_hash.to_ascii_lowercase() {
            return Err(package_error(
                "hash_mismatch",
                "artifact changed across elevation boundary",
            ));
        }
        let staged = stage_validated_archive_zip(&protected_archive, &root, transaction_id, &keys)
            .map_err(|error| package_error(error.code(), error.to_string()))?;
        activate_staged_payload_tree(&staged, &root, &keys)
            .map_err(|error| package_error(error.code(), error.to_string()))?;
        Ok(())
    })();
    let _ = std::fs::remove_dir_all(&transactions);
    activated
}

fn protected_install_root() -> Result<PathBuf, DeployError> {
    let executable = win32::module_path()?;
    let root = executable
        .parent()
        .and_then(Path::parent)
        .ok_or_else(|| package_error("privilege_boundary", "module path is unavailable"))?
        .to_path_buf();
    let program_files = win32::program_files_path()?;
    let executable_name = executable
        .file_name()
        .and_then(|value| value.to_str())
        .unwrap_or_default();
    let parent_name = executable
        .parent()
        .and_then(Path::file_name)
        .and_then(|value| value.to_str())
        .unwrap_or_default();
    if !win32::ordinal_path_prefix(&root, &program_files)?
        || !executable_name.eq_ignore_ascii_case("fcitx5-deployer.exe")
        || !parent_name.eq_ignore_ascii_case("bin")
    {
        return Err(package_error(
            "privilege_boundary",
            "deployer must run from Program Files/Fcitx5/bin",
        ));
    }
    Ok(root)
}

fn read_trusted_keys(path: impl AsRef<Path>) -> Result<Vec<TrustedKey>, DeployError> {
    let bytes = std::fs::read(path).map_err(|error| {
        package_error(
            "invalid_keyring",
            format!("trusted keyring read failed: {error}"),
        )
    })?;
    let text = std::str::from_utf8(&bytes)
        .map_err(|_| package_error("invalid_keyring", "trusted keyring is not UTF-8"))?;
    let keys =
        parse_trusted_keys(text).map_err(|error| package_error(error.code(), error.to_string()))?;
    if keys.is_empty() {
        return Err(package_error("invalid_keyring", "trusted keyring is empty"));
    }
    Ok(keys)
}

fn unicode_arg(value: &OsString) -> Result<String, DeployError> {
    value
        .to_str()
        .map(ToOwned::to_owned)
        .ok_or_else(|| internal_error("argument is not valid Unicode"))
}

fn is_hex_digest(value: &str) -> bool {
    value.len() == 64 && value.bytes().all(|byte| byte.is_ascii_hexdigit())
}

#[derive(Debug)]
enum DeployError {
    Package { code: &'static str, message: String },
    Internal(String),
}

fn package_error(code: &'static str, message: impl Into<String>) -> DeployError {
    DeployError::Package {
        code,
        message: message.into(),
    }
}

fn internal_error(message: impl Into<String>) -> DeployError {
    DeployError::Internal(message.into())
}

#[cfg(windows)]
mod win32 {
    #![allow(unsafe_code)]

    use super::{internal_error, package_error, DeployError, MAXIMUM_ARTIFACT_BYTES};
    use std::ffi::c_void;
    use std::os::windows::ffi::{OsStrExt, OsStringExt};
    use std::path::{Path, PathBuf};

    type Handle = *mut c_void;

    const INVALID_HANDLE_VALUE: Handle = -1isize as Handle;
    const TOKEN_QUERY: u32 = 0x0008;
    const TOKEN_ELEVATION_CLASS: u32 = 20;
    const GENERIC_READ: u32 = 0x8000_0000;
    const GENERIC_WRITE: u32 = 0x4000_0000;
    const OPEN_EXISTING: u32 = 3;
    const CREATE_NEW: u32 = 1;
    const FILE_ATTRIBUTE_NORMAL: u32 = 0x80;
    const FILE_FLAG_OPEN_REPARSE_POINT: u32 = 0x0020_0000;
    const FILE_FLAG_WRITE_THROUGH: u32 = 0x8000_0000;
    const FILE_ATTRIBUTE_REPARSE_POINT: u32 = 0x400;
    const FILE_ATTRIBUTE_TAG_INFO_CLASS: u32 = 9;
    const CSIDL_PROGRAM_FILES: i32 = 0x0026;
    const SHGFP_TYPE_CURRENT: u32 = 0;
    const CSTR_EQUAL: i32 = 2;

    #[repr(C)]
    struct TokenElevation {
        token_is_elevated: u32,
    }

    #[repr(C)]
    struct FileAttributeTagInfo {
        file_attributes: u32,
        reparse_tag: u32,
    }

    #[repr(C)]
    struct LargeInteger {
        quad_part: i64,
    }

    #[link(name = "shell32")]
    unsafe extern "system" {
        fn CloseHandle(object: Handle) -> i32;
        fn GetCurrentProcess() -> Handle;
        fn OpenProcessToken(process: Handle, desired_access: u32, token: *mut Handle) -> i32;
        fn GetTokenInformation(
            token: Handle,
            token_information_class: u32,
            token_information: *mut c_void,
            token_information_length: u32,
            return_length: *mut u32,
        ) -> i32;
        fn GetModuleFileNameW(module: Handle, filename: *mut u16, size: u32) -> u32;
        fn SHGetFolderPathW(
            hwnd: Handle,
            csidl: i32,
            token: Handle,
            flags: u32,
            path: *mut u16,
        ) -> i32;
        fn CompareStringOrdinal(
            string1: *const u16,
            count1: i32,
            string2: *const u16,
            count2: i32,
            ignore_case: i32,
        ) -> i32;
        fn CreateFileW(
            file_name: *const u16,
            desired_access: u32,
            share_mode: u32,
            security_attributes: Handle,
            creation_disposition: u32,
            flags_and_attributes: u32,
            template_file: Handle,
        ) -> Handle;
        fn GetFileInformationByHandleEx(
            file: Handle,
            file_information_class: u32,
            file_information: *mut c_void,
            buffer_size: u32,
        ) -> i32;
        fn GetFileSizeEx(file: Handle, file_size: *mut LargeInteger) -> i32;
        fn ReadFile(
            file: Handle,
            buffer: *mut c_void,
            bytes_to_read: u32,
            bytes_read: *mut u32,
            overlapped: Handle,
        ) -> i32;
        fn WriteFile(
            file: Handle,
            buffer: *const c_void,
            bytes_to_write: u32,
            bytes_written: *mut u32,
            overlapped: Handle,
        ) -> i32;
        fn FlushFileBuffers(file: Handle) -> i32;
    }

    struct OwnedHandle(Handle);

    impl OwnedHandle {
        fn new(handle: Handle) -> Result<Self, DeployError> {
            if handle.is_null() || handle == INVALID_HANDLE_VALUE {
                Err(package_error(
                    "artifact_changed",
                    "artifact cannot be opened exclusively",
                ))
            } else {
                Ok(Self(handle))
            }
        }

        fn raw(&self) -> Handle {
            self.0
        }
    }

    impl Drop for OwnedHandle {
        fn drop(&mut self) {
            if !self.0.is_null() && self.0 != INVALID_HANDLE_VALUE {
                unsafe {
                    let _ = CloseHandle(self.0);
                }
            }
        }
    }

    pub fn is_elevated() -> bool {
        let mut token = std::ptr::null_mut();
        let opened = unsafe { OpenProcessToken(GetCurrentProcess(), TOKEN_QUERY, &mut token) };
        let Ok(token) = OwnedHandle::new(token) else {
            return false;
        };
        if opened == 0 {
            return false;
        }
        let mut elevation = TokenElevation {
            token_is_elevated: 0,
        };
        let mut returned = 0;
        let ok = unsafe {
            GetTokenInformation(
                token.raw(),
                TOKEN_ELEVATION_CLASS,
                &mut elevation as *mut _ as *mut c_void,
                std::mem::size_of::<TokenElevation>() as u32,
                &mut returned,
            )
        };
        ok != 0 && elevation.token_is_elevated != 0
    }

    pub fn module_path() -> Result<PathBuf, DeployError> {
        let mut buffer = vec![0u16; 32768];
        let length = unsafe {
            GetModuleFileNameW(
                std::ptr::null_mut(),
                buffer.as_mut_ptr(),
                buffer.len() as u32,
            )
        };
        if length == 0 || length as usize >= buffer.len() {
            return Err(package_error(
                "privilege_boundary",
                "module path is unavailable",
            ));
        }
        buffer.truncate(length as usize);
        Ok(PathBuf::from(std::ffi::OsString::from_wide(&buffer)))
    }

    pub fn program_files_path() -> Result<PathBuf, DeployError> {
        let mut buffer = vec![0u16; 260];
        let status = unsafe {
            SHGetFolderPathW(
                std::ptr::null_mut(),
                CSIDL_PROGRAM_FILES,
                std::ptr::null_mut(),
                SHGFP_TYPE_CURRENT,
                buffer.as_mut_ptr(),
            )
        };
        if status != 0 {
            return Err(package_error(
                "privilege_boundary",
                "Program Files path is unavailable",
            ));
        }
        let len = buffer
            .iter()
            .position(|value| *value == 0)
            .unwrap_or(buffer.len());
        buffer.truncate(len);
        Ok(PathBuf::from(std::ffi::OsString::from_wide(&buffer)))
    }

    pub fn ordinal_path_prefix(candidate: &Path, parent: &Path) -> Result<bool, DeployError> {
        let candidate = std::fs::canonicalize(candidate).map_err(|error| {
            internal_error(format!("candidate path canonicalization failed: {error}"))
        })?;
        let mut parent = std::fs::canonicalize(parent).map_err(|error| {
            internal_error(format!("parent path canonicalization failed: {error}"))
        })?;
        parent.push("");
        let candidate_wide = wide_path(&candidate);
        let parent_wide = wide_path(&parent);
        if candidate_wide.len() <= parent_wide.len() {
            return Ok(false);
        }
        let result = unsafe {
            CompareStringOrdinal(
                candidate_wide.as_ptr(),
                parent_wide.len() as i32,
                parent_wide.as_ptr(),
                parent_wide.len() as i32,
                1,
            )
        };
        Ok(result == CSTR_EQUAL)
    }

    pub fn copy_exclusive_artifact(source: &Path, destination: &Path) -> Result<(), DeployError> {
        let input = OwnedHandle::new(unsafe {
            CreateFileW(
                wide_nul(source).as_ptr(),
                GENERIC_READ,
                0,
                std::ptr::null_mut(),
                OPEN_EXISTING,
                FILE_ATTRIBUTE_NORMAL | FILE_FLAG_OPEN_REPARSE_POINT,
                std::ptr::null_mut(),
            )
        })?;
        let mut tag = FileAttributeTagInfo {
            file_attributes: 0,
            reparse_tag: 0,
        };
        let mut size = LargeInteger { quad_part: 0 };
        let tag_ok = unsafe {
            GetFileInformationByHandleEx(
                input.raw(),
                FILE_ATTRIBUTE_TAG_INFO_CLASS,
                &mut tag as *mut _ as *mut c_void,
                std::mem::size_of::<FileAttributeTagInfo>() as u32,
            )
        };
        let size_ok = unsafe { GetFileSizeEx(input.raw(), &mut size) };
        if tag_ok == 0
            || size_ok == 0
            || (tag.file_attributes & FILE_ATTRIBUTE_REPARSE_POINT) != 0
            || size.quad_part <= 0
            || size.quad_part as u64 > MAXIMUM_ARTIFACT_BYTES
        {
            return Err(package_error(
                "artifact_changed",
                "artifact identity or size is unsafe",
            ));
        }
        if let Some(parent) = destination.parent() {
            std::fs::create_dir_all(parent).map_err(|error| {
                internal_error(format!("destination directory creation failed: {error}"))
            })?;
        }
        let output = OwnedHandle::new(unsafe {
            CreateFileW(
                wide_nul(destination).as_ptr(),
                GENERIC_WRITE,
                0,
                std::ptr::null_mut(),
                CREATE_NEW,
                FILE_ATTRIBUTE_NORMAL | FILE_FLAG_WRITE_THROUGH,
                std::ptr::null_mut(),
            )
        })
        .map_err(|_| package_error("io_error", "protected artifact copy cannot be created"))?;
        let mut buffer = vec![0u8; 64 * 1024];
        loop {
            let mut read = 0;
            let read_ok = unsafe {
                ReadFile(
                    input.raw(),
                    buffer.as_mut_ptr() as *mut c_void,
                    buffer.len() as u32,
                    &mut read,
                    std::ptr::null_mut(),
                )
            };
            if read_ok == 0 {
                return Err(package_error("io_error", "artifact read failed"));
            }
            if read == 0 {
                break;
            }
            let mut written = 0;
            let write_ok = unsafe {
                WriteFile(
                    output.raw(),
                    buffer.as_ptr() as *const c_void,
                    read,
                    &mut written,
                    std::ptr::null_mut(),
                )
            };
            if write_ok == 0 || written != read {
                return Err(package_error("io_error", "protected artifact copy failed"));
            }
        }
        let flushed = unsafe { FlushFileBuffers(output.raw()) };
        if flushed == 0 {
            return Err(package_error("io_error", "protected artifact flush failed"));
        }
        Ok(())
    }

    fn wide_path(path: &Path) -> Vec<u16> {
        path.as_os_str().encode_wide().collect()
    }

    fn wide_nul(path: &Path) -> Vec<u16> {
        let mut value = wide_path(path);
        value.push(0);
        value
    }
}
